src/vision/status_quo/barbara_simulates_hydrodynamics.md

```diff
@@ -0,0 +1,73 @@
+
+```rust
+    let mut stage_primitive_and_scalar = |index: BlockIndex, state: BlockState<C>, hydro: H, geometry: GridGeometry| {
+        let stage = async move {
+            let p = state.try_to_primitive(&hydro, &geometry)?;
+            let s = state.scalar_mass / &geometry.cell_volumes / p.map(P::lorentz_factor);
+            Ok::<_, HydroError>( ( p.to_shared(), s.to_shared() ) )
+        };
+        stage_map.insert(index, runtime.spawn(stage).map(|f| f.unwrap()).shared());
+    };
+```
+lacked performance because she needed to clone the value for every task.  So, Barbara switched over to using `Arc` to keep a thread safe RC to the shared data. But this change introduced a lot of `.map` and `.unwrap` function calls, making the code much harder to read. She realized that managing the dependency graph was not intuitive when using `async` for concurrency.
+
+During the move to `async` Barbara ran into a steep learning curve with error handling. The initial version of her design just used `panic!`s to fail the program if an error was encountered, but the stack traces were almost unreadable. She asked her teammate Grace to migrate over to using the `Result` idiom for error handling and Grace found a major inconvenience. The Rust type inference inconsistently breaks when propagating `Result` in `async` blocks. Grace frequently found that she had to specify the type of the error when creating a result value:
+```rust
+Ok::<_, HydroError>( ( p.to_shared(), s.to_shared() ) )  
+```
+And she could not figure out why she had to add the `::<_, HydroError>` to some of the `Result` values.
+
+Once her team began using the new `async` design for their simulations, they noticed an important issue that impacted productivity: compilation time had now increased to between 30 and 60 seconds. The nature of their work requires frequent changes to code and recompilation and 30-60 seconds is long enough to have a noticeable impact on their quality of life.  What her and her team want is for compilation to be 2 to 3 seconds. Barbara believes that the use of `async` is a major contributor to the long compilation times.
+
+This new solution works, but Barbara is not satisfied with how complex her code wound up becoming with the move to `async` and the fact that compilation time is now 30-60 seconds.  The state sharing adding a large amount of cruft with `Arc` and `async` is not well suited for using a dependency graph to schedule tasks so implementing this solution created a key component of her program that was difficult to understand and pervasive. Ultimately, her conclusion was that `async` is not appropriate for parallelizing computational tasks. She will be trying a new design based upon Rayon in the next version of her application.
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
```

Comment:
I'd like to explore the choice of `tokio` a bit more. I wonder if the choice of that particular runtime led to some of the issues (though it's clear that not all of them came from this choice). Perhaps the lack of a runtime designed for compute bound workloads is another issue here.

Reply:
That's a very good point: I'll follow up with the source to get more details on how `tokio` was chosen. Personally, I believe that this is either a missing piece of "guidance" in terms of helping people understand the type of concurrency problem they are solving (IO bound, compute bound, etc) and how to get started on the right foot for solving that type of problem.

Reply:
I don't yet know enough much about this area of the Rust ecosystem. What's the runtime refer to with regards to `tokio`?  The library as a whole, it's API, or the code under the hood that manages threads, task execution, scheduling, etc?

One big problem, from my perspective, is that the semantics of `async` are simply not able to express a complex network of thousands of compute tasks in a simple way.  Trying to write a task until 3 other tasks have completed isn't easy with `async`. In this story, Barbara had to jury rig a dependency-based scheduler into `async` that would block tasks from starting their computations until their dependencies were completed. Put another way, the semantics of `async` are simply meant for  very different type of concurrency problem than the one Barbara was solving.

@jzrake let me know if this is an accurate interpretation of your experience.

Reply:
>  this is either a missing piece of "guidance"

I 100% agree here. The async ecosystem is highly focused on server side scenarios (for the obvious reason that it's a very popular use case), but we need to do better advising users who have different needs. 

> simply meant for very different type of concurrency problem than the one Barbara was solving.

I'm not sure whether I agree or disagree with you. async/await lets you declare dependencies of a certain piece of computation in the same way you would in synchronous code:

```rust
let dep1 = dep1.await;
let dep2 = dep2.await
// now compute based on dep1 and dep2
let answer = compute(dep1, dep2);
```

If you need dependencies to run in parallel you can join them.

```rust
let deps = join_all(dep1, dep2).await;
let answer = compute(deps);
```

Obviously, the experience represents here shows that async/await as it is today does not lead to simple to understand code for compute bound workloads, but I'm just not sure if this is a fundamental property of the computation model or just lack of good guidance and good tooling for this use case.

Reply:
> I'm not sure whether I agree or disagree with you. async/await lets you declare dependencies of a certain piece of computation in the same way you would in synchronous code:
> If you need dependencies to run in parallel you can join them.
Right, but I'm wondering if that API just doesn't work well when you have thousands of dynamically created tasks that are all joined together in a complex dependency graph. Which makes me wonder if the `async` API assumes that the structure of the code defines the dependencies of the tasks.  For example, if you have tasks X, Y, Z that depend on A, B then you write:

```rust
// define tasks a and b
let deps = join_all(a, b).await;

// define tasks x, y, z
join_all(x, y, z).await;
```
and that might create some friction when trying to write code that coordinates a lot compute based tasks.

> Obviously, the experience represents here shows that async/await as it is today does not lead to simple to understand code for compute bound workloads, but I'm just not sure if this is a fundamental property of the computation model or just lack of good guidance and good tooling for this use case.
Totally agree. One of my take-aways from this story is that it's worth taking some time to see if there is a good solution to Barbara's problem using `async`, if so then we could use that to improve the documentation and guidance, and if not then ask if this is a feature that `async` is missing, or if `async` is not the right tool, etc.

Reply:
I don't believe tokio led to any issues here, even though i agree that there a futures library designed for computational workflows might help (I would think it wouldn't require deep integration into the executor).

Reply:
what I mean is: the difficulties I imagine tokio brings is perhaps that its fair scheduler could be more efficient for computation bound workflows, for example, where fairness is not a concern. But that doesn't play into this story, which is more about the challenge of picking the right thing and general "hardness" of async Rust overall. 

I do think a library that had functions or abstractions for doing computational DAGs like this might help, of course, but I don't see why that would be tied closely to the executor (at least, not yet).

Reply:
I think there are barriers to the clarity

1. there are 1000's of tasks, and linking them creates awkward code: not Tokio's fault
2. running the tasks on workers requires `spawn`: also not Tokio's fault (I think)

(1) I believe is just inherent to `async` as a construct. (2) is a problem because it couples the computation to the executor (to perform well, Tokio wants to be the spawner and the executor). If the `async` block had syntax to hint the executor it wants to run on a worker thread, such as

```rust
let y = async cpu_bound {
    calculate(x.await)
};
```

that would alleviate (2). But (1) is still an issue.

Reply:
To circle back, are there any changes to the story that I should make based on this discussion?

Reply:
After thinking about it over the weekend, I think this indicated that how I described the chain of events that led using `tokio` was not explicit enough.  So, I've update the story to make that more explicit. The first paragraph of the `Solution Path` section now has a very simple step-by-step for how the character wound up using `tokio` for compute bound parallelization.

Reply:
I'm quite happy with it.